src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -265,107 +295,122 @@ private static void CopySignatureField(byte[] signatureField, byte[] response, i
             }
         }
 
-        private void FreeKey()
+        public override int KeySize
         {
-            if (_key != null && _key.IsValueCreated)
+            get
             {
-                SafeEcKeyHandle handle = _key.Value;
-
-                if (handle != null)
+                if (_curve != null)
                 {
-                    handle.Dispose();
+                    // Load the key so we don't return 0
+                    SafeEcKeyHandle key = _key.Value;
+                    Debug.Assert(_curve == null);
                 }
+
+                return base.KeySize;
+            }
+            set
+            {
+                if (KeySize == value)
+                    return;
+
+                // Set the KeySize first so that an invalid value doesn't throw away the key
+                base.KeySize = value;
+
+                FreeKey();
+                _key = new Lazy<SafeEcKeyHandle>(GenerateKeyLazy);
             }
         }
 
-        private static int GetKeySize(SafeEcKeyHandle ecKeyHandle)
+        public override void GenerateKey(ECCurve curve)
         {
-            int nid = Interop.Crypto.EcKeyGetCurveName(ecKeyHandle);
-            int keySize = 0;
+            curve.Validate();
+            FreeKey();
 
-            for (int i = 0; i < s_supportedAlgorithms.Length; i++)
+            if (curve.IsNamed)
             {
-                if (s_supportedAlgorithms[i].Nid == nid)
-                {
-                    keySize = s_supportedAlgorithms[i].KeySize;
-                    break;
-                }
+                _curve = curve;
+                _key = new Lazy<SafeEcKeyHandle>(GenerateKeyLazy);
+                // Generate the key
+                var dummy = _key.Value;
```

Comment:
Why?

Oh, right, this is so that generate (and the ctor/static method) will throw when the curve isn't supported, right?

What feels better to me is to query the OS for the supported curves and validate it.  But, since ECC key generation is MUCH faster than RSA key generation this might be the more efficient option.

But, if that's the case, your comment should say *why* we're generating the key, not just that we're doing it.